Dockerfile

```diff
@@ -4,18 +4,36 @@ ENV VIRTUAL_ENV=/usr/src/app/venv
 RUN python -m venv $VIRTUAL_ENV
 ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 RUN pip install gunicorn
-RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python
+
+RUN curl -sSL https://install.python-poetry.org | python3 -
 COPY ./poetry.toml ./poetry.toml
 COPY ./poetry.lock ./poetry.lock
 COPY ./pyproject.toml ./pyproject.toml
-RUN /root/.poetry/bin/poetry config virtualenvs.create false \
-  && /root/.poetry/bin/poetry install
+RUN /root/.local/bin/poetry config virtualenvs.create false \
+  && /root/.local/bin/poetry install
 COPY ./todo_app ./todo_app
 
+# production build stage
 FROM base as production
 EXPOSE 5000
 CMD ["gunicorn", "--workers=2", "app:create_app()", "--bind", ":5000", "--chdir", "/usr/src/app/todo_app"]  
 
+# local development stage
 FROM base as development
 EXPOSE 5000
-CMD ["/root/.poetry/bin/poetry", "run", "flask", "run", "--host", "0.0.0.0"]
\ No newline at end of file
+CMD ["/root/.local/bin/poetry", "run", "flask", "run", "--host", "0.0.0.0"]
+
+FROM base as test
+COPY ./tests ./tests
+COPY ./tests_e2e ./tests_e2e
```

Comment:
Minor, but if you think users might run the tests in Docker locally often, then moving these copy instructions to the end of this stage would make it more common that you can take advantage of the Firefox/geckodriver installation being cached which can be quite a slow stage